tiledb/sm/array_schema/dimension.cc

```diff
@@ -59,6 +60,193 @@ class DimensionException : public StatusException {
+      const Range& mbr,
+      std::vector<uint64_t>& relevant_ranges) const override {
+    if constexpr (std::is_same_v<T, char>) {
+      assert(base.var_size());
+    }
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::relevant_ranges<T>(ranges, mbr, relevant_ranges);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  std::vector<bool> covered_vec(
+      const NDRange& ranges,
+      const Range& mbr,
+      const std::vector<uint64_t>& relevant_ranges) const override {
+    if constexpr (std::is_same_v<T, char>) {
+      assert(base.var_size());
+    }
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::covered_vec<T>(ranges, mbr, relevant_ranges);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  void split_range(const Range& r, const ByteVecValue& v, Range* r1, Range* r2)
+      const override {
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::split_range<T>(r, v, r1, r2);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  void splitting_value(
+      const Range& r, ByteVecValue* v, bool* unsplittable) const override {
+    if constexpr (std::is_same_v<T, char>) {
+      assert(base.var_size());
+    }
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::splitting_value<T>(r, v, unsplittable);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  uint64_t tile_num(const Range& range) const override {
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::tile_num<T>(&base, range);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  uint64_t map_to_uint64(
+      const void* coord,
+      uint64_t coord_size,
+      int bits,
+      uint64_t max_bucket_val) const override {
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::map_to_uint64_2<T>(
+          &base, coord, coord_size, bits, max_bucket_val);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  ByteVecValue map_from_uint64(
+      uint64_t value, int bits, uint64_t max_bucket_val) const override {
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::map_from_uint64<T>(&base, value, bits, max_bucket_val);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+  bool smaller_than(
+      const ByteVecValue& value, const Range& range) const override {
+    if constexpr (tiledb::type::TileDBFundamental<T>) {
+      return Dimension::smaller_than<T>(&base, value, range);
+    } else {
+      ::stdx::unreachable();
+    }
+  }
+};
+
+template <class T>
+class DimensionFixedSize : public DimensionDispatchTyped<T> {
+ public:
+  using DimensionDispatchTyped<T>::DimensionDispatchTyped;
+
+ protected:
+  Range compute_mbr(const WriterTile& tile) const override {
+    return Dimension::compute_mbr<T>(tile);
+  }
+
+  Range compute_mbr_var(const WriterTile&, const WriterTile&) const override {
+    ::stdx::unreachable();
```

Comment:
Definitely a `throw std::logic_error` case here.

Reply:
I haven't looked, but given that we're forcefully asserting that only one version of the mbr method is called, could we not just have one `compute_mbr` that does fixed or var based on type? Seems unnecessary to have both and then ensure the other one is never called.

Reply:
Since static_assert worked for the other functions I imagine it will work here too?
Otherwise we could probably ignore the offsets tile for the fixed-length types.

Reply:
This is definitely trickier.

So, what happened before - `set_compute_mbr_func` would check the type and then EITHER set `compute_mbr_func_ = compute_mbr<decltype(T)>`, OR set `compute_mbr_var_func_ = compute_mbr_var<char>` based on `varsize()`.

Then `Dimension::compute_mbr` asserts that `compute_mbr_func_` is set, and `Dimension::compute_mbr_var` asserts that `compute_mbr_var` is set.

So a mis-use of the API would previously have triggered an assertion failure, in identical conditions to when this `unreachable()` would occur.

IMO the change from assertion failure to `unreachable()` would be "not worse", a change to `std::logic_error` would be an improvement... and what about changing the function APIs?

There is no offset tile for fixed-length dimensions, so it would require a semantic change.

To me, any of the options is acceptable - but I'll go with `std::logic_error` and keeping the same function APIs, that seems like the most appropriate ratio of effort to reward.